src/libraries/Common/tests/Tests/System/TimeProviderTests.cs

```diff
@@ -0,0 +1,380 @@
+            mres = new ManualResetEventSlim(false);
+            ctr = token.Register(() => mres.Set());
+
+            Assert.False(token.IsCancellationRequested,
+               "CancellationTokenSourceWithTimer:  Cancellation signaled on super-long timeout (TimeSpan,1)!");
+
+            cts.CancelAfter(prettyLong);
+
+            Assert.False(token.IsCancellationRequested,
+               "CancellationTokenSourceWithTimer:  Cancellation signaled on super-long timeout (TimeSpan,2) !");
+
+            cts.CancelAfter(new TimeSpan(1000));
+
+            Debug.WriteLine("CancellationTokenSourceWithTimer: > About to wait on cancellation that should occur soon (TimeSpan)... if we hang, something bad happened");
+
+            mres.Wait();
+
+            cts.Dispose();
+        }
+
+        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
+        [MemberData(nameof(TimersProvidersListData))]
+        public static void RunDelayTests(TimeProvider provider)
+        {
+            CancellationTokenSource cts = new CancellationTokenSource();
+            CancellationToken token = cts.Token;
+
+            // These should all complete quickly, with RAN_TO_COMPLETION status.
+            Task task1 = Task.Delay(new TimeSpan(0), provider);
+            Task task2 = Task.Delay(new TimeSpan(0), provider, token);
+
+            Debug.WriteLine("RunDelayTests:    > Waiting for 0-delayed uncanceled tasks to complete.  If we hang, something went wrong.");
+            try
+            {
+                Task.WaitAll(task1, task2);
+            }
+            catch (Exception e)
+            {
+                Assert.True(false, string.Format("RunDelayTests:    > FAILED.  Unexpected exception on WaitAll(simple tasks): {0}", e));
+            }
+
+            Assert.True(task1.Status == TaskStatus.RanToCompletion, "    > FAILED.  Expected Delay(TimeSpan(0), timeProvider) to run to completion");
+            Assert.True(task2.Status == TaskStatus.RanToCompletion, "    > FAILED.  Expected Delay(TimeSpan(0), timeProvider, uncanceledToken) to run to completion");
+
+            // This should take some time
+            Task task3 = Task.Delay(TimeSpan.FromMilliseconds(20000), provider);
+            Assert.False(task3.IsCompleted, "RunDelayTests:    > FAILED.  Delay(20000) appears to have completed too soon(1).");
+            Task t2 = Task.Delay(TimeSpan.FromMilliseconds(10));
+            Assert.False(task3.IsCompleted, "RunDelayTests:    > FAILED.  Delay(10000) appears to have completed too soon(2).");
+        }
+
+        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
+        [MemberData(nameof(TimersProvidersListData))]
+        public static async void RunWaitAsyncTests(TimeProvider provider)
+        {
+            CancellationTokenSource cts = new CancellationTokenSource();
+
+            Task task = Task.Run(() => { while (true) {} });
```

Comment:
These tasks will never finish, even after the test finishes, and will continue burning a core for the remainder of the test process.  These tasks should be stopped when you no longer need them, and the test should wait on them to make sure the work has quiesced before the test returns.

Reply:
Oooops :-( I didn't mean to leave it like that.